src/ast/nodes.rs

```diff
@@ -0,0 +1,443 @@
+            }
+        }
+
+        impl rowan::ast::AstNode for $name {
+            type Language = crate::NixLanguage;
+
+            fn can_cast(kind: crate::SyntaxKind) -> bool {
+                kind == $kind
+            }
+
+            fn cast(from: SyntaxNode) -> Option<Self> {
+                if Self::can_cast(from.kind()) {
+                    Some(Self(from))
+                } else {
+                    None
+                }
+            }
+            fn syntax(&self) -> &SyntaxNode {
+                &self.0
+            }
+        }
+
+        ast_nodes!{ @impl $name $($tt)* }
+    };
+
+    ({ $($kind:ident => $typed:ident),* $(,)? } => $name:ident $($tt:tt)*) => {
+        pub enum $name {
+            $(
+                $typed($typed),
+            )*
+        }
+
+        impl std::fmt::Display for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                std::fmt::Display::fmt(self.syntax(), f)
+            }
+        }
+
+        impl rowan::ast::AstNode for $name {
+            type Language = NixLanguage;
+
+            fn can_cast(kind: SyntaxKind) -> bool {
+                matches!(kind, $($kind)|*)
+            }
+
+            fn cast(syntax: SyntaxNode) -> Option<Self> {
+                let res = match syntax.kind() {
+                    $(
+                        $kind => $name::$typed($typed(syntax))
+                    ),*,
+                    _ => return None,
+                };
+                Some(res)
+            }
+
+            fn syntax(&self) -> &SyntaxNode {
+                match self {
+                    $(
+                        $name::$typed(it) => &it.0,
+                    )*
+                }
+            }
+        }
+
+        ast_nodes! { @impl $name $($tt)* }
+    };
+
+    (@impl $name:ident $(: $trait:ident)* $(: { $($item:item)* })?, $($tt:tt)*) => {
+        $(impl $name {
+            $($item)*
+        })?
+
+        $(impl $trait for $name {})*
+
+        ast_nodes! { $($tt)* }
+    };
+
+    () => { };
+}
+
+ast_nodes! {
+     {
+        NODE_APPLY => Apply,
+        NODE_ASSERT => Assert,
+        NODE_DYNAMIC => Dynamic,
+        NODE_ERROR => Error,
+        NODE_IF_ELSE => IfElse,
+        NODE_SELECT => Select,
+        NODE_INHERIT => Inherit,
+        NODE_INHERIT_FROM => InheritFrom,
+        NODE_LITERAL => Literal,
+        NODE_LAMBDA => Lambda,
+        NODE_LEGACY_LET => LegacyLet,
+        NODE_LET_IN => LetIn,
+        NODE_LIST => List,
+        NODE_BIN_OP => BinOp,
+        NODE_OR_DEFAULT => OrDefault,
+        NODE_PAREN => Paren,
+        NODE_ROOT => Root,
+        NODE_ATTR_SET => AttrSet,
+        NODE_UNARY_OP => UnaryOp,
+        NODE_IDENT => Ident,
+        NODE_WITH => With,
+    } => Expr,
+
+    NODE_LITERAL => Literal,
+
+    {
+        NODE_DYNAMIC => Dynamic,
+        NODE_STRING => Str,
+    } => Key,
+
+     NODE_IDENT => Ident: {
```

Comment:
There are quite some indentation errors inside this macro. Please reformat them.
```suggestion
    NODE_IDENT => Ident: {
```